env/env_encryption.cc

```diff
@@ -361,9 +413,152 @@ Status EncryptedRandomRWFile::Close() { return file_->Close(); }
+      // Initialize & write prefix (if needed)
+      AlignedBuffer buffer;
+      Slice prefix;
+      *prefix_length = provider->GetPrefixLength();
+      if (*prefix_length > 0) {
+        // Initialize prefix
+        buffer.Alignment(underlying->GetRequiredBufferAlignment());
+        buffer.AllocateNewBuffer(*prefix_length);
+        status = provider->CreateNewPrefix(fname, buffer.BufferStart(),
+                                           *prefix_length);
+        if (status.ok()) {
+          buffer.Size(*prefix_length);
+          prefix = Slice(buffer.BufferStart(), buffer.CurrentSize());
+          // Write prefix
+          status = underlying->Append(prefix);
+        }
+        if (!status.ok()) {
+          return status;
+        }
+      }
+      // Create cipher stream
+      status = provider->CreateCipherStream(fname, options, prefix, stream);
+    }
+    return status;
+  }
+
+  template <class TypeFile>
+  Status CreateRandomWriteCipherStream(
+      const std::string& fname, const std::unique_ptr<TypeFile>& underlying,
+      const EnvOptions& options, size_t* prefix_length,
+      std::unique_ptr<BlockAccessCipherStream>* stream) {
+    EncryptionProvider* provider = nullptr;
+    *prefix_length = 0;
+    Status status = GetWritableProvider(fname, &provider);
+    if (!status.ok()) {
+      return status;
+    } else if (provider != nullptr) {
+      // Initialize & write prefix (if needed)
+      AlignedBuffer buffer;
+      Slice prefix;
+      *prefix_length = provider->GetPrefixLength();
+      if (*prefix_length > 0) {
+        // Initialize prefix
+        buffer.Alignment(underlying->GetRequiredBufferAlignment());
+        buffer.AllocateNewBuffer(*prefix_length);
+        status = provider->CreateNewPrefix(fname, buffer.BufferStart(),
+                                           *prefix_length);
+        if (status.ok()) {
+          buffer.Size(*prefix_length);
+          prefix = Slice(buffer.BufferStart(), buffer.CurrentSize());
+          // Write prefix
+          status = underlying->Write(0, prefix);
+        }
+        if (!status.ok()) {
+          return status;
+        }
+      }
+      // Create cipher stream
+      status = provider->CreateCipherStream(fname, options, prefix, stream);
+    }
+    return status;
+  }
+
+  template <class TypeFile>
+  Status CreateSequentialCipherStream(
+      const std::string& fname, const std::unique_ptr<TypeFile>& underlying,
+      const EnvOptions& options, size_t* prefix_length,
+      std::unique_ptr<BlockAccessCipherStream>* stream) {
+    // Read prefix (if needed)
+    AlignedBuffer buffer;
+    Slice prefix;
+    *prefix_length = provider_->GetPrefixLength();
+    if (*prefix_length > 0) {
+      // Read prefix
+      buffer.Alignment(underlying->GetRequiredBufferAlignment());
+      buffer.AllocateNewBuffer(*prefix_length);
+      Status status =
+          underlying->Read(*prefix_length, &prefix, buffer.BufferStart());
```

Comment:
if this were a PositionedRead, the file offset does not move.  Then failure of CreateCipherStream because the file is not encrypted "just works".  If CreateCipherStream succeeds, the underlying->Skip(*prefix_length) to get offset aligned with next read point.

Reply:
Define a specific status code for CreateCipherStream to return if not encrypted ... either a new kNotEncrypted or maybe kAborted (pick one, I do not care).  Then we need a BlockAccessCipherStream specifically for not encrypted files ... that is what every provider returns as a not-me equivalent.

Reply:
*prefix_length gets set to zero in that case ... fixing that part of the equation.

Reply:
status = CreateSequentialCipherStream(fname, underlying, options,
                                          &prefix_length, &stream);
    if (status.ok()) {
      result->reset(new EncryptedSequentialFile(
          std::move(underlying), std::move(stream), prefix_length));
    }
    return status;

becomes

    status = CreateRandomReadCipherStream(fname, underlying, options,
                                          &prefix_length, &stream);
    if (status.ok()) {
      underlying->Skip(prefix_length);
      result->reset(new EncryptedRandomAccessFile(
          std::move(underlying), std::move(stream), prefix_length));
    } else if (status.IsNotEncrypted() {
        status = status.OK();
    }